packages/admin-ui/src/toast/use-toast-queue-state.ts

```diff
@@ -0,0 +1,100 @@
+import { useReducer, useCallback } from 'react'
+import type { InternalToastProps } from './types'
+
+export function useToastQueueState() {
+  const [{ toasts }, dispatch] = useReducer(queueReducer, {
+    toasts: [],
+    queue: {},
+  })
+
+  const add = useCallback(
+    (toast: InternalToastProps) => dispatch({ type: 'enqueue', toast }),
+    []
+  )
+
+  const remove = useCallback(
+    (key: string) => dispatch({ type: 'dequeue', key }),
+    []
+  )
+
+  return {
+    toasts,
+    add,
+    remove,
+  }
+}
+
+function queueReducer(
+  state: ToastQueueState,
+  action: QueueActions
+): ToastQueueState {
+  switch (action.type) {
+    case 'enqueue': {
+      const { toast } = action
+
+      const hasExistingToast = state.toasts.some(
+        (t) => t.dedupeKey === toast.dedupeKey
+      )
+
+      if (hasExistingToast) {
+        const { dedupeKey } = toast
+
+        return {
+          toasts: state.toasts.map((t) => {
+            if (t.dedupeKey === dedupeKey) {
+              return {
+                ...t,
+                shouldRemove: true,
+              }
+            }
+
+            return t
```

Comment:
Suggestion: Just to help the readability inside the map function. Use it if it makes sense to you

```suggestion
          toasts: state.toasts.map((t) =>
              t.dedupeKey === dedupeKey ? { ...t, shouldRemove: true } : t
          )
```